lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,217 @@
+            # NOTE: sample's timestamp represent the end of sampling
+            # so each sample actually starts at the timestamp of previous sample
+
+            for timestamp, raw_sample in self._raw_samples[device][
+                1:
+            ]:  # ignore first empty sample
+                params, current = self._split_output(raw_sample)
+                current = self._parse_values(current)
+
+                packets = int(current["sofar"]) - packets_sofar
+
+                samples[device].append(
+                    {
+                        "timestamp": start_timestamp,
+                        "duration": timestamp - start_timestamp,
+                        "packets": packets,
+                        "errors": int(current["errors"]),
+                        "params": params,
+                    }
+                )
+                packets_sofar += packets
+                start_timestamp = timestamp
+
+        return samples
+
+    def _read_dev_outputs(self) -> Iterator[tuple[str, str]]:
+        for device in self._devs:
+            output = ""
+            with open(f"/proc/net/pktgen/{device}", "r") as f:
+                output = f.readlines()
+            yield device, "\n".join(output)
+
+    def _split_output(self, output: str) -> tuple:
+        match = re.search(r"Params:(.+)Current:(.+)Result:\s(?:\w+)", output, re.DOTALL)
+        if not match:
+            raise TestModuleError(f"Could not parse pktgen devide output: {output}")
+
+        return match.groups()
+
+    def _parse_values(self, params) -> dict[str, str]:
+        values = {}
+
+        for key, value in re.findall(r"(\w+):?\s(\S+)", params, re.MULTILINE):
+            values[key.lower()] = value
+
+        return values
+
+
+class PktGen(BaseTestModule):
+    """
+    In the scope of this module, the physical interface is refered as `interface`.
+    Pktgen device (interface@anything) is refered as device.
+
+    Inspired by https://github.com/torvalds/linux/blob/master/samples/pktgen/pktgen_sample03_burst_single_flow.sh
+    """
+
+    cpus = ListParam(type=IntParam())  # each CPU is 1 generator
+
+    src_if = DeviceParam()
+    dst_mac = StrParam()
+
+    src_ip = IpParam()
+    dst_ip = IpParam()
+
+    count = IntParam(default=0)  # 0 = no upper limit
+    pkt_size = IntParam(default=60)  # 4 bytes are added for CRC by NIC
+    frags = IntParam(default=1)
+    burst = IntParam(default=8)
+
+    duration = IntParam(default=60)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self._devices = []
+
+        self._res_data = {}
+        self._output_parser = None
+
+    def run(self):
+        if not kmod_loaded("pktgen"):
+            raise TestModuleError("pktgen module is not loaded")
+
+        self._pg_ctrl("reset")
+        self._configure_generator()
+
+        output_parser = PktGenResultsParser(self._devices, self.params.duration)
```

Comment:
I got confused by the class name here. The name `PktGenResultsParser` evokes that this should just parse the results (which it does), however the next line says `start_sampling`. Could we just rename it to `PktGenResultsSampler`?

Then it would look like:
```
sampler = PktGenResultsSampler(self._devices, self.params.duration)
sampler.start_sampling()
...
self._res_data = sampler.parse_dev_outputs()
```

The change above would make this more readable imo.

Additionally when I look at the proposed code above, can't it be simply:
```
self._res_data = sampler.dev_samples # rename + switch method to property? that would call _parse_dev_outputs() as a private method?
```

What do you think?

Reply:
Sure, thanks